nixos/modules/services/databases/neo4j.nix

```diff
@@ -36,114 +89,541 @@ let
-      environment = {
-        NEO4J_HOME = "${cfg.package}/share/neo4j";
-        NEO4J_CONF = "${cfg.dataDir}/conf";
-      };
-      serviceConfig = {
-        ExecStart = "${cfg.package}/bin/neo4j console";
-        User = "neo4j";
-        PermissionsStartOnly = true;
-        LimitNOFILE = 40000;
-      };
-      preStart = ''
-        mkdir -m 0700 -p ${cfg.dataDir}/{data/graph.db,conf,logs}
-        ln -fs ${serverConfig} ${cfg.dataDir}/conf/neo4j.conf
-        if [ "$(id -u)" = 0 ]; then chown -R neo4j ${cfg.dataDir}; fi
-      '';
-    };
+  config =
+    let
+      policyNameList = (attrNames cfg.ssl.policies);
+      validPolicyNameList = ["legacy"] ++ policyNameList;
+      validPolicyNameString = concatStringsSep ", " validPolicyNameList;
+    in
+
+    mkIf cfg.enable {
+      assertions = [
+        { assertion = !(elem "legacy" policyNameList);
+          message = "The policy 'legacy' is special to Neo4j, and its name is reserved."; }
+        { assertion = elem cfg.bolt.sslPolicy validPolicyNameList;
+          message = "Invalid policy assigned: `services.neo4j.bolt.sslPolicy = ${cfg.bolt.sslPolicy}`, defined policies are: ${validPolicyNameString}"; }
+        { assertion = elem cfg.https.sslPolicy validPolicyNameList;
+          message = "Invalid policy assigned: `services.neo4j.https.sslPolicy = ${cfg.https.sslPolicy}`, defined policies are: ${validPolicyNameString}"; }
+      ];
+
+      systemd.services.neo4j = {
+        description = "Neo4j Daemon";
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" ];
+        environment = {
+          NEO4J_HOME = "${cfg.package}/share/neo4j";
+          NEO4J_CONF = "${toString cfg.directories.home}/conf";
+        };
+        serviceConfig = {
+          ExecStart = "${cfg.package}/bin/neo4j console";
+          User = "neo4j";
+          PermissionsStartOnly = true;
+          LimitNOFILE = 40000;
+        };
+
+        preStart = ''
+          # Directories Setup
+          mkdir -m 0700 -p ${toString cfg.directories.home}/{certificates,conf,data,import,logs,plugins}
 
-    environment.systemPackages = [ cfg.package ];
+          # SSL Policies Directory setup
+          #   Only generate required directories when default certificates and baseDirectory is used.
+          ${concatStringsSep "\n" (mapAttrsToList (name: value:
+            if cfg.directories.home + "/certificates/" + name == value.baseDirectory then
+              "mkdir -m 0700 -p ${toString value.baseDirectory}/{revoked,trusted}"
+            else "")
+          (cfg.ssl.policies))}
 
-    users.users = singleton {
-      name = "neo4j";
-      uid = config.ids.uids.neo4j;
-      description = "Neo4j daemon user";
-      home = cfg.dataDir;
+          # Place the configuration where Neo4j can find it.
+          ln -fs ${serverConfig} ${toString cfg.directories.home}/conf/neo4j.conf
+
+          # Ensure neo4j user ownership
+          chown -R neo4j ${toString cfg.directories.home}
```

Comment:
I think it would be better to remove all `toString` functions from all those directories. How it is now would make `certificates = ./path/to/certs` behave as if you specified `"/absolute/path/to/certs"`, but what one would expect (just from what all other modules do) is that ./path/to/certs gets imported into the store and result in `"/nix/store/<hash>-certs"`. `toString` can still be applied from the user if it's what they need like this.